File: tennis-intelligence-platform-web/src/types/match.ts
// types/match.ts — matches api/schemas/match.py exactly (MatchReplayResponse,
// MatchSearchResponse). Field names and nullability are taken directly from the
// real, confirmed API responses returned by the running FastAPI backend, not
// guessed — see api/schemas/match.py for the Pydantic source of truth.

export interface PlayerRef {
  name: string;
}

export interface PrematchProbabilities {
  markov: number | null;
  ml_mc: number | null;
  ml_informed_unsmoothed: number | null;
  ml_informed_smoothed: number | null;
  hybrid: number | null;
}

export interface PointPrediction {
  point_index: number;
  set1: number;
  set2: number;
  gm1: number;
  gm2: number;
  markov_p1: number;
  ml_mc_p1: number;
  ml_informed_unsmoothed_p1: number;
  ml_informed_smoothed_p1: number;
  hybrid_p1: number;
}

export interface SetBoundary {
  set_number: number;
  point_index: number;
  score: string;
  winner_is_p1: boolean;
}

export interface MatchReplayResponse {
  match_id: string;
  player1: PlayerRef;
  player2: PlayerRef;
  winner: string;
  n_points: number;
  tournament: string | null;
  date: string | null;
  final_score: string | null;
  prematch: PrematchProbabilities;
  points: PointPrediction[];
  set_boundaries: SetBoundary[];
}

export interface MatchSearchResponse {
  match_ids: string[];
}

// The five prediction engines, used consistently across the UI for labels/colors —
// matches this project's own ENGINE_DISPLAY_NAMES convention (replay_service.py,
// export_model_comparison.py) rather than inventing separate frontend naming.
export type EngineKey = "markov" | "ml_mc" | "ml_informed_unsmoothed" | "ml_informed_smoothed" | "hybrid";

export const ENGINE_LABELS: Record<EngineKey, string> = {
  markov: "Analytical Markov",
  ml_mc: "Machine Learning + Monte Carlo",
  ml_informed_unsmoothed: "ML-Informed Markov (Unsmoothed)",
  ml_informed_smoothed: "ML-Informed Markov (Smoothed)",
  hybrid: "Hybrid Engine",
};

// Maps each engine to its point-level field name in PointPrediction — since the
// real API uses a per-engine suffix convention (markov_p1, ml_mc_p1, etc.) rather
// than a nested object, this lookup avoids repeating that mapping in every
// component that needs to pull a specific engine's trajectory.
export const ENGINE_POINT_FIELD: Record<EngineKey, keyof PointPrediction> = {
  markov: "markov_p1",
  ml_mc: "ml_mc_p1",
  ml_informed_unsmoothed: "ml_informed_unsmoothed_p1",
  ml_informed_smoothed: "ml_informed_smoothed_p1",
  hybrid: "hybrid_p1",
};

File: tennis-intelligence-platform/api/schemas/match.py
"""
schemas/match.py — Pydantic response models for match-related endpoints.

Field names and shapes here are DELIBERATELY kept identical to what
tennis_intel.serving.replay_service.replay_match_by_id() already returns as a plain
dict — these models exist to give FastAPI request validation and OpenAPI docs, not to
introduce a second, divergent shape the frontend would need to reconcile.
"""

from __future__ import annotations

from pydantic import BaseModel


class PlayerRef(BaseModel):
    name: str


class PrematchProbabilities(BaseModel):
    """One or more of these fields is null when that engine has no genuine,
    zero-information pre-match computation available in this pipeline (see
    replay_match.py's own extensive comment on this exact point) — the frontend
    should render a dash, not a fabricated 50%, for any null field here."""
    markov: float | None = None
    ml_mc: float | None = None
    ml_informed_unsmoothed: float | None = None
    ml_informed_smoothed: float | None = None
    hybrid: float | None = None


class PointPrediction(BaseModel):
    point_index: int
    set1: int
    set2: int
    gm1: int
    gm2: int
    markov_p1: float
    ml_mc_p1: float
    ml_informed_unsmoothed_p1: float
    ml_informed_smoothed_p1: float
    hybrid_p1: float


class SetBoundary(BaseModel):
    set_number: int
    point_index: int
    score: str
    winner_is_p1: bool


class MatchReplayResponse(BaseModel):
    match_id: str
    player1: PlayerRef
    player2: PlayerRef
    winner: str
    n_points: int
    tournament: str | None = None
    date: str | None = None
    final_score: str | None = None
    prematch: PrematchProbabilities
    points: list[PointPrediction]
    set_boundaries: list[SetBoundary]


class MatchSearchResponse(BaseModel):
    match_ids: list[str]